File: 2130.cpp
#include <algorithm>
#include <deque>
#include <iostream>
#include <map>
#include <set>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
  int pairSum(ListNode *head) {
    ListNode *f = head;
    ListNode *s = head;
    while (f && f->next) {
      s = s->next;
      f = f->next->next;
    }
    ListNode *prev = nullptr;
    ListNode *next = nullptr;
    while (s) {
      next = s->next;
      s->next = prev;
      prev = s;
      s = next;
    }
    int ans = INT32_MIN;
    while (prev && head) {
      ans = max(ans, prev->val + head->val);
      prev = prev->next;
      head = head->next;
    }
    return ans;
  }
};
int main() {}

File: 2130.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# Cases:
# x
# 1 -> 2
# 1 -> 2 -> 3 -> 4

# reverse the 2nd half of the list and iterate forwards.


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        f = head
        s = head
        while f and f.next:
            f = f.next.next
            s = s.next
        prev = None
        next = None
        while s:
            next = s.next
            s.next = prev
            prev = s
            s = next
        head2 = prev
        ans = -float("inf")
        while head2 and head:
            ans = max(head2.val + head.val, ans)
            head2 = head2.next
            head = head.next
        return ans
